validator/src/bootstrap.rs

```diff
@@ -387,159 +514,112 @@ pub fn rpc_bootstrap(
-
-                if expected_genesis_hash != rpc_genesis_hash {
-                    return Err(format!(
-                        "Genesis hash mismatch: expected {} but RPC node genesis hash is {}",
-                        expected_genesis_hash, rpc_genesis_hash
-                    ));
-                }
-            }
-
-            let (cluster_info, gossip_exit_flag, gossip_service) = gossip.take().unwrap();
-            cluster_info.save_contact_info();
-            gossip_exit_flag.store(true, Ordering::Relaxed);
-            gossip_service.join().unwrap();
-
-            let rpc_client_slot = rpc_client
-                .get_slot_with_commitment(CommitmentConfig::finalized())
-                .map_err(|err| format!("Failed to get RPC node slot: {}", err))?;
-            info!("RPC node root slot: {}", rpc_client_slot);
-
-            download_snapshots(
-                full_snapshot_archives_dir,
-                incremental_snapshot_archives_dir,
+        while vetted_rpc_nodes.is_empty() {
+            let rpc_node_details_vec = get_rpc_nodes(
+                &gossip.as_ref().unwrap().0,
+                cluster_entrypoints,
                 validator_config,
+                &mut blacklisted_rpc_nodes.write().unwrap(),
                 &bootstrap_config,
-                use_progress_bar,
-                maximum_local_snapshot_age,
-                start_progress,
-                minimal_snapshot_download_speed,
-                maximum_snapshot_download_abort,
-                &mut download_abort_count,
-                snapshot_hash,
-                &rpc_contact_info,
-            )
-        })
-        .map(|_| {
-            if let Some(url) = bootstrap_config.check_vote_account.as_ref() {
-                let rpc_client = RpcClient::new(url);
-                check_vote_account(
-                    &rpc_client,
-                    &identity_keypair.pubkey(),
-                    vote_account,
-                    &authorized_voter_keypairs
-                        .read()
-                        .unwrap()
-                        .iter()
-                        .map(|k| k.pubkey())
-                        .collect::<Vec<_>>(),
-                )
-                .unwrap_or_else(|err| {
-                    // Consider failures here to be more likely due to user error (eg,
-                    // incorrect `solana-validator` command-line arguments) rather than the
-                    // RPC node failing.
-                    //
-                    // Power users can always use the `--no-check-vote-account` option to
-                    // bypass this check entirely
-                    error!("{}", err);
-                    exit(1);
-                });
+            );
+            if rpc_node_details_vec.is_empty() {
+                return;
             }
-        });
 
-        if result.is_ok() {
-            break;
+            vetted_rpc_nodes = rpc_node_details_vec
+                .into_par_iter()
```

Comment:
hmm, rayon is generally not ideal for an IO-bound problem, but I guess if it works.

Reply:
Yeah, this doesn't help w/ the actual download speed. Just speeds up discovery of viable RPC nodes through:
1. Attempting to connect to a bunch of them at once (this is where thread concurrency helps)
2. Only waiting 5 seconds for timeout instead of 30 seconds (default timeout)

Reply:
Yea, I got that. My point was the work time in the rayon thread will be generally waiting for a network response, it won't be on computing something locally with the CPU. Like doing a math-heavy calculation with multiple cores.

This is another explanation:
https://users.rust-lang.org/t/parrallel-io-data-processing-fork-join-thread-spawn-or-rayon-or-other/64687/4

But there's not much happening in the validator at this moment, so ultimate efficiency is probably not that important.

Reply:
Ah, I see. Yeah, async w/ Tokyo would probably be better here, but as you mentioned, there's nothing else going on in the system at this point